src/schema/search.ts

```diff
@@ -0,0 +1,148 @@
+import {
+  GraphQLList,
+  GraphQLEnumType,
+  GraphQLString,
+  GraphQLNonNull,
+  GraphQLFieldConfig,
+} from "graphql"
+import { connectionFromArraySlice } from "graphql-relay"
+import { connectionWithCursorInfo } from "schema/fields/pagination"
+import { pageable } from "relay-cursor-paging"
+import { createPageCursors } from "schema/fields/pagination"
+import { SearchableItem } from "./searchableItem"
+
+import { Searchable } from "./searchable"
+import { convertConnectionArgsToGravityArgs } from "lib/helpers"
+
+import { visit } from "graphql"
+
+export const SearchIndex = new GraphQLEnumType({
+  name: "SearchIndexes",
+  values: {
+    ARTWORK: {
+      value: "Artwork",
+    },
+    ARTIST: {
+      value: "Artist",
+    },
+  },
+})
+
+export const SearchMode = new GraphQLEnumType({
+  name: "SearchMode",
+  values: {
+    AUTOSUGGEST: {
+      value: "AUTOSUGGEST",
+    },
+    SITE: {
+      value: "SITE",
+    },
+  },
+})
+
+export const searchArgs = pageable({
+  query: {
+    type: new GraphQLNonNull(GraphQLString),
+    description: "Search query to perform. Required.",
+  },
+  indexes: {
+    type: new GraphQLList(SearchIndex),
+    description: "Indexes to include in search. Default: [ARTIST, ARTWORK].",
+  },
+  mode: {
+    type: SearchMode,
+    description: "Mode of search to exceute. Default: SITE.",
+  },
+})
+
+const hydrateSearchResult = (result, root) => {
+  const loaderMapping = {
+    Artist: root.artistLoader,
+    Artwork: root.artworkLoader,
+  }
+
+  const loader = loaderMapping[result.label]
+
+  return loader(result.id).then(object => {
+    return {
+      ...object,
+      __typename: result.label,
+    }
+  })
+}
+
+const SearchConnection = connectionWithCursorInfo(Searchable)
+
+export const Search: GraphQLFieldConfig<any, any, any> = {
+  type: SearchConnection,
+  description: "Global search",
+  args: searchArgs,
+  resolve: (source, args, _request, info) => {
+    const pageOptions = convertConnectionArgsToGravityArgs(args)
+
+    let shouldHydrate = false
+
+    visit(info.fieldNodes[0], {
+      Field(node) {
+        if (node.name.value === "node") {
+          visit(node, {
+            InlineFragment(node) {
+              if (
+                node.typeCondition &&
+                node.typeCondition.name.value !== SearchableItem.name
+              ) {
+                shouldHydrate = true
```

Comment:
We can probably return the BREAK constant here to stop further recursion of the AST